pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,471 @@
+				); failure != nil {
+					failures = append(failures, *failure...)
+				}
+			}
+		}
+	// The number and string cases are for enums. The normal cases are handled
+	// in the matchType function right now we are not validating enums. Enums
+	// are more of a helper, but are not exhaustive in all cases i.e. ec2
+	// instance types
+	case "string":
+	case "number":
+	}
+
+	if len(failures) > 0 {
+		return &failures
+	}
+
+	return nil
+}
+
+// getType gets a type definition from a schema reference. Currently it only
+// supports types from the same schema
+func (v *PulumiInputValidator) getType(typeRef string) *pschema.ComplexTypeSpec {
+	if strings.HasPrefix(typeRef, "#/types/") {
+		ref := strings.TrimPrefix(typeRef, "#/types/")
+		if typeSpec, ok := v.schema.Types[ref]; ok {
+			return &typeSpec
+		}
+	}
+	return nil
+}
+
+// ValidateInputs will validate a set of inputs against the pulumi schema. It will
+// return a list of type failures if any are found
+func (v *PulumiInputValidator) ValidateInputs(inputs resource.PropertyMap) *[]TypeFailure {
+	failures := []TypeFailure{}
+	for key, value := range inputs {
+		if failure := v.validateInputType(string(key), value, pathBuilder{paths: []string{}}); failure != nil {
+			failures = append(failures, *failure...)
+		}
+	}
+
+	if len(failures) > 0 {
+		return &failures
+	}
+
+	return nil
+}
+
+func (v *PulumiInputValidator) validateInputType(
+	inputName string,
+	inputValue resource.PropertyValue,
+	pathBuilder pathBuilder,
+) *[]TypeFailure {
+	resourceType := v.urn.Type()
+	failures := []TypeFailure{}
+	if resource, ok := v.schema.Resources[resourceType.String()]; ok {
+		if prop, ok := resource.InputProperties[inputName]; ok {
+			if failure := v.validateTypeSpec(
+				inputName,
+				inputValue,
+				prop.TypeSpec,
+				pathBuilder,
+				resource.RequiredInputs,
+			); failure != nil {
+				failures = append(failures, *failure...)
+			}
+		} else {
+			return &[]TypeFailure{
+				{
+					ResourcePath: pathBuilder.toPath(),
+					Reason:       fmt.Sprintf("property %s is not defined in the schema", inputName),
+				},
+			}
+		}
+	}
+
+	if len(failures) > 0 {
+		return &failures
+	}
+
+	return nil
+}
+
+type possibleTypes []string
+
+func (p possibleTypes) add(typ ...string) possibleTypes {
+	for _, t := range typ {
+		if !slices.Contains(p, t) && t != "" {
+			p = append(p, t)
+
+		}
+	}
+	return p
+}
+
+func (p possibleTypes) toString() string {
+	return strings.Join(p, " OR ")
+}
+
+// findOneOf attempts to find the correct type spec for a given input input value
+// from a list of possible types
+func (v *PulumiInputValidator) findOneOf(
+	inputValue resource.PropertyValue,
+	specs []pschema.TypeSpec,
+) (*pschema.TypeSpec, []string, bool) {
+
+	allSpecs := []string{}
+	for _, spec := range specs {
+		allSpecs = append(allSpecs, spec.Type)
+		switch inputValue.TypeString() {
+		case "[]":
+			if spec.Type == "array" {
+				if spec.Items != nil {
+					return spec.Items, []string{}, true
```

Comment:
This jumps from an array type like `Array<string>` to an element type like `string`, which feels very strange and unrelated to the idea of unrolling OneOf. The entire function I'd like to understand what it's trying to do, starting from the signature as it does not seem to use inputValue.

Reply:
I was trying to brush up on OneOf semantic, here's a good piece of code that interprets it during "binding" to Type.

```https://github.com/pulumi/pulumi/blob/master/pkg/codegen/schema/bind.go#L842```

This is actually interesting as it leads to UnionType https://github.com/pulumi/pulumi/blob/master/pkg/codegen/schema/schema.go#L185

It's interesting -- this code might be easier to write over the bound Type objects than raw TypeSpec

Also thinking that the schema is optimized more for code generation (primary purpose) than type checking we're attempting here, so interpreting UnionType.DefaultType is a little tricky.

IN the first pass it's not unreasonable to punt on union types entirely.

If we were to attempt to type check them, I imagine something could be tried along the lines of:

```
var ut UnionType
var foundCleanMatch
for i, possibleType := range ut.ElementTypes {
   tcRes = tc(value, possibleType)
   if !tcRes.Error() {
      foundCleanMatch = true
   }
}
if foundCleanMatch {
   return result.NoError()
}
return result.Format("Unexpected value %v, expecting one of %s", value, formatOptions(..))
```

This can also get tedious to get right from usability pov as deeply nested union errors can get very verbose.. But we're looking for something to try out here for the first time so no need to get it perfect.

Reply:
I added a comment to this function to try and better explain what it is doing.


 For example, given the following schema:
```
	"prop": {
		TypeSpec: pschema.TypeSpec{
			OneOf: []pschema.TypeSpec{
				{
					Type: "array",
					Items: &pschema.TypeSpec{
						Type: "object",
						// using ref to test specific object keys
						Ref: "#/types/pkg:index/type:ObjectStringType",
					},
				},
				{
					Type: "string",
				},
			},
		},
	}
```

 and the following input value:
```
	"prop": []map[string]interface{}{{"key": "value"}}
```
 In this case the actual type that we want to compare "prop" to is the `ObjectStringType`
 so this function will return the `Items` `TypeSpec` for comparison.

Reply:
This doesn't make sense to me tbh. Let's dig in here on a call.

Reply:
based on our conversation I think I was able to refactor this to what we discussed. Let me know what you think.